src/result/result.py

```diff
@@ -537,3 +539,67 @@ def do(gen: Generator[Result[T, E], None, None]) -> Result[T, E]:
     except DoException as e:
         out: Err[E] = e.err  # type: ignore
         return out
+    except TypeError as te:
+        # Turn this into a more helpful error message.
+        # Python has strange rules involving turning generators involving `await`
+        # into async generators, so we want to make sure to help the user clearly.
+        if "'async_generator' object is not an iterator" in str(te):
+            raise TypeError(
+                "You have hit a case where `do()` does not support async calls.\n"
+                "Either use do_async(), or move your `await` expressions before `do()`\n."
+                "Example:\n"
+                "foo = await bar\ndo(...for foo_ok in foo)\n"
+            )
+        raise te
+
+
+async def do_async(
+    gen: Union[Generator[Result[T, E], None, None], AsyncGenerator[Result[T, E], None]]
+) -> Result[T, E]:
+    """Async version of do. Example:
+
+    >>> final_result: Result[float, int] = await do_async(
+        Ok(len(x) + int(y) + z)
+            for x in await get_async_result_1()
+            for y in await get_async_result_2()
+            for z in get_sync_result_3()
+        )
+
+    NOTE: Python makes generators async in a counter-intuitive way.
+    This is a regular generator:
+        async def foo(): ...
+        do(Ok(1) for x in await foo())
+
+    But this is an async generator:
+        async def foo(): ...
+        async def bar(): ...
+        do(
+            Ok(1)
+            for x in await foo()
+            for y in await bar()
+        )
+
+    We let users try to use regular `do()`, which works in some cases
+    of awaiting async values. If we hit a case like above, we raise
+    an exception telling the user to use `do_async()` instead.
+    See `do()`.
+
+    However, for better usability, it's better for `do_async()` to also accept
+    regular generators, as you get in the first case:
+
+    async def foo(): ...
+        do(Ok(1) for x in await foo())
+
+    Furthermore, neither mypy nor pyright can infer that the second case is
+    actually an async generator, so we cannot annotate `do_async()`
+    as accepting only an async generator. This is additional motivation
+    to accept either.
+    """
+    try:
+        if isinstance(gen, AsyncGenerator):
+            return await gen.__anext__()
+        else:
+            return next(gen)
+    except DoException as e:
+        out: Err[E] = e.err  # type: ignore
+        return out
```

Comment:
Why don't we just replace `do` with this implementation? This seems to cover both cases, I don't see an obvious need to have both. Just having the one that works in all cases seems like the better choice.

Aside from this extra `isinstance` check at runtime,

Reply:
Yeah interesting idea. You really need the two "front-ends" because one is async and has to be `await`ed. It would be weird to have to await every do call even if you're not using any actual asyncio.

Question is how much implementation they can share, and I think the answer is not much. If you look at them side by side, they're pretty different. I could abstract out the `try...  except DoException` but I think that's really overdoing abstraction. The duplicated code amounts to maybe 4 lines.

Reply:
You're right, I didn't consider that Python necessitates awaits for async stuff and it would look weird for non-async.